### PR TITLE
planner: Stabilise TestPartitionWithVariedDataSources

### LINKED ISSUE #36740: unstable test TestPartitionWithVariedDataSources
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
[2022-07-30T14:02:41.681Z]         	Error:      	Not equal: 

[2022-07-30T14:02:41.681Z]         	            	expected: "[1]\n"

[2022-07-30T14:02:41.681Z]         	            	actual  : "[0]\n"

[2022-07-30T14:02:41.681Z]         	            	

[2022-07-30T14:02:41.681Z]         	            	Diff:

[2022-07-30T14:02:41.681Z]         	            	--- Expected

[2022-07-30T14:02:41.681Z]         	            	+++ Actual

[2022-07-30T14:02:41.681Z]         	            	@@ -1,2 +1,2 @@

[2022-07-30T14:02:41.681Z]         	            	-[1]

[2022-07-30T14:02:41.681Z]         	            	+[0]

[2022-07-30T14:02:41.681Z]         	            	 

[2022-07-30T14:02:41.681Z]         	Test:       	TestPartitionWithVariedDataSources

[2022-07-30T14:02:41.681Z]         	Messages:   	sql:select @@last_plan_from_cache /* table: tnormalIdx */, args:[]

```

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

Issue Number: close #36740

Problem Summary:
Checking if a query was served from the plan cache or not may be unstable, when doing it 100s of times.
So allow a small fraction to not be cached, when the vast majority is.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by mjonss** [comment]
I ran `(cd planner/core ; go test -v -run TestPartitionWithVariedDataSources -count 1000 -failfast)` in 18 parallel runs and they all timed out after 10 min, each running 70+ times, while without this change it would fail in multiple cases. So I think it is stable now.

**Review by qw4990 — APPROVED** [review]

File: planner/core/prepare_test.go [hunk 1]
```diff
@@ -2758,6 +2775,11 @@ func TestPartitionWithVariedDataSources(t *testing.T) {
 			}
 		}
 	}
+	// Allow ~1% non-cached queries, due to background changes etc.
+	// (Actually just 1/3 %, since there are 3 tables * 4 queries per loop :)
+	if len(missedPlanCache) > (loops * 4 / 100) {
```
Comment (xhebox) [inline planner/core/prepare_test.go:2780]: `len(missedPlanCache)` looks like we only need a simple integer counter?

Comment (mjonss) [reply planner/core/prepare_test.go:2780]: Yes, but using the `[]string` array, we can also show all the different tables/plans it failed, so I think the combined use is ok.

**Review by xhebox — APPROVED** [review]

**Pull request merged**